### PR TITLE
Show candidate hosts for the given query

### PR DESCRIPTION
**Pull request by navis**
Provide location information for the given query. Need for calculating location of input split.

### CODE REVIEW

**Comment by fjy** [comment]
@navis can you provide some more information on why this is feature is required. You can get this information through Druid metrics.

**Comment by navis** [comment]
@fjy I'm thinking of implementing a druid storage handler for hive. marked as 'wip'.

**Comment by himanshug** [comment]
@navis having a hive storage adapter would be good but it appears you are thinking of featching data from historical nodes (and realtime nodes). IMO it will be much more scalable to read segments from hdfs directly.
I wrote a hadoop InputFormat ( a simple wrapper around druid's DatasourceInputFormat) and a Pig loader that works that way (planning to move InputFormat to druid core) , see

**Comment by navis** [comment]
Because we are supposed to use both of druid(simple select/groupby) and hive(for joins, udfs, etc) for the same dataset, it's highly possible that historical node already loaded segments to be used and we want to reuse it if possible. I think we can add a method to ask the historical node about local temporary directory it used for bypassing druid access. 

I don't know how the InputFormat you've mentioned is implemented. Is it support predicates to be pushed to use bitmap index in druid smooth file? If it is, there is not much work to be done for me :).

**Comment by himanshug** [comment]
@navis the InputFormat I refered to is https://github.com/himanshug/druid-hadoop-utils/blob/master/druid-mr/src/main/java/com/yahoo/druid/hadoop/DruidInputFormat.java which is a simple wrapper on https://github.com/druid-io/druid/blob/master/indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceInputFormat.java .

Regarding pushing predicates and iterating through rows in the segment, you can do that with https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/segment/realtime/firehose/IngestSegmentFirehose.java which is used in https://github.com/druid-io/druid/blob/master/indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceRecordReader.java .

**Comment by navis** [comment]
@himanshug Thanks, I'll check that.

**Comment by navis** [comment]
Anyway, below is the test result in my notebook.

```
navis@navisui-MacBook-Pro:~/druid$ explain_query.sh timeseries.json
[ {
  "itvl" : "2016-01-28T09:00:00.000+09:00/2016-01-29T09:00:00.000+09:00",
  "ver" : "2016-01-28T23:47:17.046Z",
  "part" : 0,
  "locations" : [ {
    "name" : "localhost:8083",
    "host" : "localhost:8083",
    "maxSize" : 10000000000,
    "type" : "historical",
    "tier" : "_default_tier",
    "priority" : 0
  } ]
}, {
  "itvl" : "2016-02-02T09:00:00.000+09:00/2016-02-03T09:00:00.000+09:00",
  "ver" : "2016-02-02T01:03:07.127Z",
  "part" : 0,
  "locations" : [ {
    "name" : "localhost:8100",
    "host" : "localhost:8100",
    "maxSize" : 0,
    "type" : "realtime",
    "tier" : "_default_tier",
    "priority" : 0
  } ]
} ]
```

**Comment by navis** [comment]
Anyway, below is the test result in my notebook.

```
navis@navisui-MacBook-Pro:~/druid$ explain_query.sh timeseries.json
[ {
  "itvl" : "2016-01-28T09:00:00.000+09:00/2016-01-29T09:00:00.000+09:00",
  "ver" : "2016-01-28T23:47:17.046Z",
  "part" : 0,
  "locations" : [ {
    "name" : "localhost:8083",
    "host" : "localhost:8083",
    "maxSize" : 10000000000,
    "type" : "historical",
    "tier" : "_default_tier",
    "priority" : 0
  } ]
}, {
  "itvl" : "2016-02-02T09:00:00.000+09:00/2016-02-03T09:00:00.000+09:00",
  "ver" : "2016-02-02T01:03:07.127Z",
  "part" : 0,
  "locations" : [ {
    "name" : "localhost:8100",
    "host" : "localhost:8100",
    "maxSize" : 0,
    "type" : "realtime",
    "tier" : "_default_tier",
    "priority" : 0
  } ]
} ]
```

**Comment by navis** [comment]
Added get method something like `curl  "http://localhost:8082/druid/v2/candidates/?datasource=wikipedia&intervals=2013-08-31/2020-09-01"`.

@himanshug I need know target segments and locations from predicates (or sarg) made in hive. `DataSourceInputFormat` is a good start point but still need this PR to cover our use case.

**Comment by navis** [comment]
I've made patch integrating druid with hive based on this patch. Can anyone review this? I think this is fairly simple patch.

**Comment by binlijin** [comment]
Looks like the coordinator also have the segment's info, so what time is the good to query coordinator and what time is the goot to query broker?

File: server/src/main/java/io/druid/server/BrokerQueryResource.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  {
+    TimelineLookup<String, ServerSelector> timeline = brokerServerView.getTimeline(datasource);
+    if (timeline == null) {
+      return Collections.emptyList();
+    }
+    List<LocatedSegmentDescriptor> located = Lists.newArrayList();
+    for (Interval interval : intervals) {
+      for (TimelineObjectHolder<String, ServerSelector> holder : timeline.lookup(interval)) {
+        for (PartitionChunk<ServerSelector> chunk : holder.getObject()) {
+          ServerSelector selector = chunk.getObject();
```
Comment (binlijin) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: Size can get from ServerSelector?

File: common/src/main/java/io/druid/timeline/TimelineObjectHolder.java [hunk 2]
```diff
@@ -28,16 +28,19 @@
 {
   private final Interval interval;
   private final VersionType version;
+  private final long approxSize;
   private final PartitionHolder<ObjectType> object;
 
   public TimelineObjectHolder(
       Interval interval,
       VersionType version,
+      long approxSize,
```
Comment (xvrl) [inline common/src/main/java/io/druid/timeline/TimelineObjectHolder.java]: do we have to maintain this value as part of the timeline object? Memory pressure is already very high on the brokers from maintaining the serverview and timeline. Is there a way we could get this information when requested as opposed to storing it all the time?

File: server/src/main/java/io/druid/server/BrokerQueryResource.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          final SegmentDescriptor descriptor = new SegmentDescriptor(
+              holder.getInterval(), holder.getVersion(), chunk.getChunkNumber()
+          );
+          List<DruidServerMetadata> candidates = selector.getCandidates();
+          located.add(new LocatedSegmentDescriptor(descriptor, holder.getApproximatedSize(), candidates));
```
Comment (xvrl) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: can you explain what approximate size represents and what it's purpose is?

File: server/src/main/java/io/druid/server/BrokerQueryResource.java (hunk 1 above)
Comment (navis) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: I cannot understand how I've missed that three months ago. Then we can remove controversial `approxSize` part safely. Thanks.

File: common/src/main/java/io/druid/timeline/TimelineObjectHolder.java (hunk 2 above)
Comment (navis) [inline common/src/main/java/io/druid/timeline/TimelineObjectHolder.java]: removed that part as binlijin said.

File: server/src/main/java/io/druid/server/BrokerQueryResource.java (hunk 3 above)
Comment (navis) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: that's for combining splits for user provided threshold, which is like things done in `DatasourceInputFormat`.

**Comment by binlijin** [comment]
Patch looks much clean now.

**Comment by binlijin** [comment]
Need doc  for how to use it.

**Comment by navis** [comment]
Fail of `AnnouncerTest.testSanity`, which seemed definitely not related to this.

**Comment by xvrl** [comment]
@navis this is useful, we built something similar standalone for our #2330 prototype, but we'll use this going forward since I think it makes sense for this to live on a broker.

File: server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ */
+public class LocatedSegmentDescriptor extends SegmentDescriptor
+{
+  private final long size;
+  private final List<DruidServerMetadata> locations;
+
+  @JsonCreator
+  public LocatedSegmentDescriptor(
+      @JsonProperty("itvl") Interval interval,
```
Comment (xvrl) [inline server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java]: for public APIs we should probably use fully qualified names instead of abbreviations.

File: server/src/main/java/io/druid/server/BrokerQueryResource.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  @Path("/candidates")
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  @Consumes({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE, APPLICATION_SMILE})
+  public Response getQueryTargets(
+      InputStream in,
+      @QueryParam("pretty") String pretty,
+      @Context final HttpServletRequest req // used only to get request content-type and remote address
+  ) throws IOException
+  {
+    final String reqContentType = req.getContentType();
```
Comment (xvrl) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: there's a lot of cut and pate from QueryResource here, can we refactor this a bit?

File: docs/content/design/broker.md [hunk 6]
```diff
@@ -63,3 +63,12 @@ Returns the metrics of the datasource.
 
 Returns a flag indicating if the broker knows about all segments in Zookeeper. This can be used to know when a broker node is ready to be queried after a restart.
 
+* `/druid/v2/candidates/?datasource=<datasource-name>&intervals=<comma-separated-intervals-in-ISO8601-format`
```
Comment (xvrl) [inline docs/content/design/broker.md]: Our APIs are not very consistent between the coordinator and the broker here
Coordinator returns almost identical information at `/druid/coordinator/v1/datasources/{dataSourceName}/intervals/{interval}/serverview` but uses a very different API

File: server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java (hunk 4 above)
Comment (navis) [inline server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java]: good.

File: server/src/main/java/io/druid/server/BrokerQueryResource.java (hunk 5 above)
Comment (navis) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: tried couple of times but it didn't make code any cleaner because it's complex code accessing many parameters. I'll try once more with shorter grip.

File: docs/content/design/broker.md (hunk 6 above)
Comment (navis) [inline docs/content/design/broker.md]: Any preference? I can make in either way.

File: server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java (hunk 4 above)
Comment (xvrl) [inline server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java]: SegmentDescriptor is not exposed to the user so it's less of a problem. However, all our classes that are exposed to the user, use either `interval` or `intervals` so we should keep that consistent.

File: docs/content/design/broker.md [hunk 7]
```diff
@@ -63,3 +63,12 @@ Returns the metrics of the datasource.
...
+
+Returns segment information lists for the given datasource and intervals.
+
+### PUT
```
Comment (xvrl) [inline docs/content/design/broker.md]: I think you meant POST

File: docs/content/design/broker.md [hunk 8]
```diff
@@ -63,3 +63,12 @@ Returns the metrics of the datasource.
...
+
+* `/druid/v2/candidates/`
```
Comment (xvrl) [inline docs/content/design/broker.md]: another possibility, which might reduce code duplication between the two endpoints would be to make this a query context flag, but that might make the QueryResources code more complex.

File: server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java (hunk 4 above)
Comment (navis) [inline server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java]: good.

File: docs/content/design/broker.md (hunk 7 above)
Comment (navis) [inline docs/content/design/broker.md]: ;)

File: docs/content/design/broker.md (hunk 8 above)
Comment (navis) [inline docs/content/design/broker.md]: then historical or realtime node can provide the information, too. I think it's possible. Would I?

File: docs/content/design/broker.md (hunk 6 above)
Comment (navis) [inline docs/content/design/broker.md]: addressed comment. now it's `/druid/v2/datasources/{dataSourceName}/intervals/{comma-separated-intervals-in-ISO8601-format}`

File: server/src/main/java/io/druid/client/selector/ServerSelector.java [hunk 9]
```diff
@@ -78,6 +81,16 @@ public boolean isEmpty()
     }
   }
 
+  public List<DruidServerMetadata> getCandidates() {
```
Comment (b-slim) [inline server/src/main/java/io/druid/client/selector/ServerSelector.java]: @navis i am wondering if we need to include the Tier logic ?

File: server/src/main/java/io/druid/server/BrokerQueryResource.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      return context.ok(
+          ServerViewUtil.getTargetLocations(
+              brokerServerView,
+              query.getDataSource(),
+              query.getIntervals()
+          )
+      );
+    }
+    catch (Exception e) {
+      return context.gotError(e);
```
Comment (b-slim) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: can we sanitize this like ```java
LOG.error(e, "Error handling request");
      return Response.serverError().entity(ServletResourceUtils.sanitizeException(e)).build();
```

File: server/src/main/java/io/druid/server/QueryResource.java [hunk 11]
```diff
@@ -394,16 +384,56 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
...
          .emit();
 
+      return context.gotError(e);
+    }
+    finally {
+      Thread.currentThread().setName(currThreadName);
+    }
+  }
+
+  protected class RequestContext
```
Comment (b-slim) [inline server/src/main/java/io/druid/server/QueryResource.java]: @navis this is a good generic way to handle context, it is not related to this PR thought, can we make this a separate PR i guess it can be part of the common utils ?

**Comment by b-slim** [comment]
@navis thanks for this PR. i guess this is an important thing to have within druid ASAP ! would love to see more UTs and maybe logic about Tiers in druid (or we can add this later and call it out in the docs).

File: server/src/main/java/io/druid/client/selector/ServerSelector.java (hunk 9 above)
Comment (navis) [inline server/src/main/java/io/druid/client/selector/ServerSelector.java]: changed to use tier selector but not sure on it.

File: server/src/main/java/io/druid/server/BrokerQueryResource.java (hunk 10 above)
Comment (navis) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: wrapped with QueryInterruptedException if needed

File: server/src/main/java/io/druid/server/QueryResource.java (hunk 11 above)
Comment (navis) [inline server/src/main/java/io/druid/server/QueryResource.java]: tried but there were some missing libraries in common module. changed to static class, anyway.

File: server/src/main/java/io/druid/client/selector/ServerSelector.java [hunk 12]
```diff
@@ -78,20 +83,45 @@ public boolean isEmpty()
...
+
+  public QueryableDruidServer pick()
+  {
+    synchronized (this) {
+      return strategy.pick(toServers(), segment.get());
+    }
+  }
 
-      return strategy.pick(prioritizedServers, segment.get());
+  private TreeMap<Integer, Set<QueryableDruidServer>> toServers()
```
Comment (b-slim) [inline server/src/main/java/io/druid/client/selector/ServerSelector.java]: minor: but i would call this `toPrioritizedServers`

File: server/src/main/java/io/druid/client/selector/ServerSelector.java [hunk 13]
```diff
@@ -78,20 +83,45 @@ public boolean isEmpty()
...
-          theServers = Sets.newHashSet();
-          prioritizedServers.put(server.getServer().getPriority(), theServers);
+      final DataSegment target = segment.get();
+      for (Map.Entry<Integer, Set<QueryableDruidServer>> entry : toServers().entrySet()) {
+        Set<QueryableDruidServer> servers = entry.getValue();
+        TreeMap<Integer, Set<QueryableDruidServer>> tieredMap = Maps.newTreeMap();
+        while (!servers.isEmpty()) {
+          tieredMap.put(entry.getKey(), servers);   // strategy.pick() removes entry
+          QueryableDruidServer server = strategy.pick(tieredMap, target);
+          result.add(server.getServer().getMetadata());
```
Comment (b-slim) [inline server/src/main/java/io/druid/client/selector/ServerSelector.java]: i guess we need a null check here ?

File: server/src/main/java/io/druid/server/BrokerQueryResource.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+      AuthConfig authConfig,
+      TimelineServerView brokerServerView
+  )
+  {
+    super(warehouse, config, jsonMapper, smileMapper, texasRanger, emitter, requestLogger, queryManager, authConfig);
+    this.brokerServerView = brokerServerView;
+  }
+
+  @POST
+  @Path("/candidates")
```
Comment (b-slim) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: i guess after https://github.com/druid-io/druid/pull/2424 we need to add authConfig logic here ?

File: server/src/main/java/io/druid/server/ClientInfoResource.java [hunk 15]
```diff
@@ -300,6 +303,24 @@ public int compare(Interval o1, Interval o2)
     return metrics;
   }
 
+  @GET
+  @Path("/{dataSourceName}/candidates/intervals/{intervals}/numCandidates/{numCandidates}")
```
Comment (b-slim) [inline server/src/main/java/io/druid/server/ClientInfoResource.java]: i guess after the changed in https://github.com/druid-io/druid/pull/2424 by @pjain1 we need to add Auth logic ?

File: server/src/main/java/io/druid/client/ServerViewUtil.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.Collections;
+import java.util.List;
+
+/**
+ */
+public class ServerViewUtil
+{
+  public static final int DEFAULT_NUM_CANDIDATES = 5;
+
+  public static List<LocatedSegmentDescriptor> getTargetLocations(
```
Comment (b-slim) [inline server/src/main/java/io/druid/client/ServerViewUtil.java]: can we add couple of UTs to this core method ?

File: server/src/main/java/io/druid/client/ServerViewUtil.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+import org.joda.time.Interval;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ */
+public class ServerViewUtil
+{
+  public static final int DEFAULT_NUM_CANDIDATES = 5;
```
Comment (b-slim) [inline server/src/main/java/io/druid/client/ServerViewUtil.java]: minor: @navis someone my want to have all the server by default but maybe be you have a specific use case in mind ?

File: docs/content/design/broker.md [hunk 18]
```diff
@@ -59,7 +59,17 @@ Returns the dimensions of the datasource.
 
 Returns the metrics of the datasource.
 
+* `/druid/v2/datasources/{dataSourceName}/candidates/intervals/{comma-separated-intervals-in-ISO8601-format}/numCandidates/{numCandidates}`
+
+Returns segment information lists including server locations for the given datasource and intervals.
```
Comment (b-slim) [inline docs/content/design/broker.md]: if we elect to go with a short list of 5 server if `numCanditates` is absent, we should add this to the doc i guess.

File: server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+public class LocatedSegmentDescriptor
+{
+  private final Interval interval;
+  private final String version;
+  private final int partitionNumber;
+  private final long size;
+  private final List<DruidServerMetadata> locations;
+
+  @JsonCreator
+  public LocatedSegmentDescriptor(
```
Comment (b-slim) [inline server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java]: some ser/desr test will be nice thought.

**Comment by b-slim** [comment]
@navis i meant some not same sorry !

File: server/src/main/java/io/druid/server/ClientInfoResource.java (hunk 15 above)
Comment (pjain1) [inline server/src/main/java/io/druid/server/ClientInfoResource.java]: Adding `@ResourceFilters(DatasourceResourceFilter.class)` to the endpoint will add the auth logic. So the endpoint will look like this -

```
  @GET
  @Path("/{dataSourceName}/candidates/intervals/{intervals}/numCandidates/{numCandidates}")
  @Produces(MediaType.APPLICATION_JSON)
  @ResourceFilters(DatasourceResourceFilter.class)
  public Iterable<LocatedSegmentDescriptor> getQueryTargets(
      @PathParam("dataSourceName") String datasource,
      @PathParam("intervals") String intervals,
      @PathParam("numCandidates") String numCandidates,
      @Context final HttpServletRequest req
  ) throws IOException
```
If this can be added then it's great otherwise I can do a follow up PR to add it.

File: server/src/main/java/io/druid/server/BrokerQueryResource.java (hunk 14 above)
Comment (pjain1) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: Adding `@ResourceFilters(StateResourceFilter.class)` to the endpoint should work. So, the endpoint will look like - 
```
  @POST
  @Path("/candidates")
  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
  @Consumes({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE, APPLICATION_SMILE})
  @ResourceFilters(StateResourceFilter.class)
  public Response getQueryTargets(
      InputStream in,
      @QueryParam("pretty") String pretty,
      @QueryParam("numCandidates") String numCandidates,
      @Context final HttpServletRequest req
  ) throws IOException
```
If this can be added then it's great otherwise I can do a follow up PR to add it. Thanks

Comment (pjain1) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: Also this resource class needs to be added to the unit tests like this `getRequestPaths(BrokerQueryResource.class)` here - https://github.com/druid-io/druid/blob/master/server/src/test/java/io/druid/server/http/security/SecurityResourceFilterTest.java#L70

File: server/src/main/java/io/druid/client/selector/ServerSelector.java (hunk 12 above)
Comment (navis) [inline server/src/main/java/io/druid/client/selector/ServerSelector.java]: ok

File: server/src/main/java/io/druid/client/selector/ServerSelector.java (hunk 13 above)
Comment (navis) [inline server/src/main/java/io/druid/client/selector/ServerSelector.java]: It does not return `null` in current implementations, but ok.

File: server/src/main/java/io/druid/server/BrokerQueryResource.java (hunk 14 above)
Comment (navis) [inline server/src/main/java/io/druid/server/BrokerQueryResource.java]: I don't know well on security things. @pjain1 Could you do that in following PR?

File: server/src/main/java/io/druid/client/ServerViewUtil.java (hunk 16 above)
Comment (navis) [inline server/src/main/java/io/druid/client/ServerViewUtil.java]: done

File: server/src/main/java/io/druid/client/ServerViewUtil.java (hunk 17 above)
Comment (navis) [inline server/src/main/java/io/druid/client/ServerViewUtil.java]: not really. I'll change default behavior to get all servers.

File: docs/content/design/broker.md (hunk 18 above)
Comment (navis) [inline docs/content/design/broker.md]: now default returns all servers. updated doc anyway.

File: server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java (hunk 19 above)
Comment (navis) [inline server/src/main/java/io/druid/query/LocatedSegmentDescriptor.java]: done

**Comment by navis** [comment]
@b-slim Addressed comments

File: server/src/main/java/io/druid/server/ClientInfoResource.java [hunk 20]
```diff
@@ -300,6 +303,24 @@ public int compare(Interval o1, Interval o2)
...
+      @PathParam("dataSourceName") String datasource,
+      @PathParam("intervals") String intervals,
+      @PathParam("numCandidates") String numCandidates,
+      @Context final HttpServletRequest req
+  ) throws IOException
+  {
+    List<Interval> intervalList = Lists.newArrayList();
+    for (String interval : intervals.split(",")) {
+      intervalList.add(Interval.parse(interval.trim()));
+    }
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/server/ClientInfoResource.java]: When Intervals are specified as PathParam do we need to do any handling for '/' ? 
e.g similar thing is done in DataSourcesResource.deleteDataSourceSpecificInterval

**Comment by nishantmonu51** [comment]
looks good to me after https://github.com/druid-io/druid/pull/2282#discussion_r79800599

File: server/src/main/java/io/druid/server/ClientInfoResource.java (hunk 20 above)
Comment (b-slim) [inline server/src/main/java/io/druid/server/ClientInfoResource.java]: i guess we need to have something like this https://github.com/druid-io/druid/blob/0d745ee1207df4990ab92082ec33c75f5777757e/server/src/main/java/io/druid/server/http/DatasourcesResource.java#L346

Comment (navis) [inline server/src/main/java/io/druid/server/ClientInfoResource.java]: oh, good

Comment (navis) [inline server/src/main/java/io/druid/server/ClientInfoResource.java]: changed to query-param for intervals/numCandidates. it looks better than before, I think.

**Comment by navis** [comment]
failed `io.druid.server.lookup.PollingLookupTest` which seemed not related to this.

**Comment by drcrallen** [comment]
@navis https://github.com/druid-io/druid/issues/3480

**Comment by nishantmonu51** [comment]
closing & reopening for travis

**Comment by nishantmonu51** [comment]
All review comments seem handled, changed milestone to 0.9.2.

**Comment by b-slim** [comment]
@pjain1 this is in can you do the follow up PR for 0.9.2 ?
